### PR TITLE
Sync Responds With Canonical Block Lists

### PR DESCRIPTION
**Pull request by terencechain**
Instead of returning batched blocks using slots, we should respond the canonical block list by tracking what sync responded to peer ID's for finalized root and head root. Using those 2 roots we can recursively construct the canonical block list

I also added a feature flag `SyncProvideCanonicalList`, it's disabled for default

### CODE REVIEW

File: beacon-chain/sync/regular_sync.go [hunk 1]
```diff
@@ -555,18 +571,34 @@ func (rs *RegularSync) handleBatchedBlockRequest(msg p2p.Message) error {
...
 	}
 
-	response := make([]*pb.BeaconBlock, 0, blockRange)
-	for i := startSlot; i <= endSlot; i++ {
-		retBlock, err := rs.db.BlockBySlot(ctx, i)
+	var response []*pb.BeaconBlock
+	if featureconfig.FeatureConfig().SyncProvideCanonicalList {
+		var peerID [32]byte
+		copy(peerID[:], msg.Peer)
+		canonicalListReq := rs.chainHeadReqMap[peerID]
```
Comment (rauljordan) [inline beacon-chain/sync/regular_sync.go]: Needs to check if exists in map

File: shared/featureconfig/config.go [hunk 2]
```diff
@@ -33,6 +33,7 @@ type FeatureFlagConfig struct {
 	DisableGossipSub              bool // DisableGossipSub in p2p messaging.
 	EnableCommitteesCache         bool // EnableCommitteesCache for state transition.
 	CacheTreeHash                 bool // CacheTreeHash determent whether tree hashes will be cached.
+	SyncProvideCanonicalList      bool // SyncProvideCanonicalList determines how sync service responds blocks by building canonical list.
```
Comment (rauljordan) [inline shared/featureconfig/config.go]: Remove

File: beacon-chain/sync/regular_sync_test.go [hunk 3]
```diff
@@ -741,3 +743,98 @@ func TestHandleStateReq_OK(t *testing.T) {
...
 }
+
+func TestCanonicalBlockList_CanRetrieveCanonical(t *testing.T) {
+	db := internal.SetupDB(t)
+	defer internal.TeardownDB(t, db)
+	ss := setupService(db)
+
+	// Construct the following chain:
+	//    /- B3
+	//	 B1  - B2 - B4
```
Comment (prestonvanloon) [inline beacon-chain/sync/regular_sync_test.go]: I think this is messed up?

Comment (rauljordan) [reply beacon-chain/sync/regular_sync_test.go]: It looks good in IDE, might be github messing it up

File: proto/beacon/p2p/v1/messages.proto [hunk 4]
```diff
@@ -46,8 +46,8 @@ message BeaconBlockResponse {
 }
 
 message BatchedBeaconBlockRequest {
```
Comment (prestonvanloon) [inline proto/beacon/p2p/v1/messages.proto]:
```suggestion
message BatchedBeaconBlockRequest {
  uint64 start_slot = 1 [deprecated=true];
  uint64 end_slot = 2 [deprecated=true];
  bytes finalized_root = 3;
  bytes canonical_root = 4;
}
```

**Pull request merged**